### PR TITLE
[Merged by Bors] - Non blocking warmup

### PR DESCRIPTION
**Pull request by almogdepaz**
## Motivation
blocks grpc server causing usabilty issues

Closes #

## Changes
cache warm up runs async

## Test Plan
same

### CODE REVIEW

File: mesh/meshdb.go [hunk 1]
```diff
@@ -714,10 +716,12 @@ func (m *MeshDB) Retrieve(key []byte, v interface{}) (interface{}, error) {
 	return v, nil
 }
 
-func (m *MeshDB) CacheWarmUp(from types.LayerID, to types.LayerID) error {
+func (m *MeshDB) cacheWarmUpFromTo(from types.LayerID, to types.LayerID) error {
 	m.Info("warming up cache with layers %v to %v", from, to)
 	for i := from; i < to; i++ {
-
+		if m.exit {
```
Comment (y0sher) [inline mesh/meshdb.go]: this is not going to work, as exit is not protected and used concurrently.

File: mesh/meshdb.go [hunk 2]
```diff
@@ -714,10 +716,12 @@ func (m *MeshDB) Retrieve(key []byte, v interface{}) (interface{}, error) {
...
+			m.Info("shutdown during cache warm up")
```
Comment (y0sher) [inline mesh/meshdb.go:725]: if shutting down why not return or break here ?

File: cmd/node/node.go [hunk 3]
```diff
@@ -501,7 +501,7 @@ func (app *SpacemeshApp) initServices(nodeID types.NodeId,
 	if mdb.PersistentData() {
 		trtl = tortoise.NewRecoveredAlgorithm(mdb, app.addLogger(TrtlLogger, lg))
 		msh = mesh.NewRecoveredMesh(mdb, atxdb, app.Config.REWARD, trtl, app.txPool, atxpool, processor, app.addLogger(MeshLogger, lg))
-
+		go msh.CacheWarmUp()
```
Comment (y0sher) [inline cmd/node/node.go:504]: I'm out of context so I don't know why we need that to be non blocking. so I'm not saying anything about that. though for this purpose I understand why you moved it from inside `NewRecoveredMesh` to here...

**Review by y0sher — APPROVED** [review]
Approved as most of my comments are minor. though I suggest you address them before merging :)

File: cmd/node/node.go (hunk 3 above)
Comment (almogdepaz) [reply cmd/node/node.go:504]: blocks the startup till warmup is done, since the app waits for the grpc server

File: mesh/meshdb.go (hunk 1 above)
Comment (almogdepaz) [reply mesh/meshdb.go]: sure il make it atomic ot something

Comment (almogdepaz) [reply mesh/meshdb.go]: thanks, supposed to be a chan.

File: mesh/meshdb.go (hunk 2 above)
Comment (almogdepaz) [reply mesh/meshdb.go:725]: missing return statement

**Comment by almogdepaz** [comment]
bors merge

**Pull request closed without merging**